### ISSUE TITLE
Feature Request: Add a method for reseting a form's child element validations for form reuse

### ISSUE THREAD

**Issue by dspigarelli** [kaliber5/ember-bootstrap#331]
I have a `bs-form` embedded in a `bs-modal`.  If any of the form's child elements have been focused on during a previous iteration of the modal, the child element retains the `showOwnValidation=true` and immediately display their validation when the modal is opened.  The request is some way to reset the `showOwnValidation` back to `false` between iterations of the modal.

Work around for now is to surround the `bs-modal` or embedded `bs-form` with an `{{#if...}}` conditioned on the modal's display.

Example:

```
{{#if showModal}}
  {{#bs-modal open=showModal ...}}
    {{#bs-form}}
    ...
    {{/bs-form}}
  {{/bs-modal}}
{{/if}}
```

---

Output from `ember bootstrap:info` (available since 1.0.0-alpha.9):
```
 $ ember bootstrap:info
Npm packages:
ember-bootstrap: 1.0.0-alpha.12 -> 1.0.0-alpha.12
ember-cli: 2.13.2 -> 2.13.2
bootstrap: n/a
bootstrap-sass: ^3.3.7 -> 3.3.7
ember-cli-sass: 6.1.3 -> 6.1.3
ember-cli-less: n/a
Bower packages:
bootstrap: n/a
bootstrap-sass: n/a
ember-bootstrap configuration:
importBootstrapCSS: false
```

**Comment by rwwagner90** [comment]
@simonihmig how can I help push this forward?

**Comment by jelhan** [comment]
@rwwagner90 Great to see that you are willing to work on this one. I think discussing public API would be the first step.

The title of this issue assumes that it should be done per form element. But the use case requires per form. We should get an agreement of what to support.

Some of the components provided by ember bootstrap use an argument to reset internal state. Setting the property bound to it, resets the internal state. E.g. `<BsButton>` has a `@reset` argument, which resets the fulfilled / rejected state.

To be honest I'm not a big fan of that pattern. But I'm also not aware of an alternative pattern. And consistency of the APIs is also an important point.

For this specific case I'm wondering if it should be integrated with `@model` argument. Maybe validation state should be reset when object used as model changes? 🤔

**Comment by rwwagner90** [comment]
@jelhan 

> For this specific case I'm wondering if it should be integrated with @model argument. Maybe validation state should be reset when object used as model changes?

This sounds reasonable. I think the model for our forms changes each time. The problem is the form elements themselves do not always change, so if one form has the same field as another and we switch forms, we end up with stale validation state.

**Comment by jelhan** [comment]
> @jelhan
> 
> > For this specific case I'm wondering if it should be integrated with @model argument. Maybe validation state should be reset when object used as model changes?
> 
> This sounds reasonable. I think the model for our forms changes each time. The problem is the form elements themselves do not always change, so if one form has the same field as another and we switch forms, we end up with stale validation state.

Not sure if I can follow. I wasn't describing a feature existing today but a potential public API to serve the use case.

To be more precise I think we would need to add an additional argument `@resetValidationStateOnModelChange` to `<BsForm>`. If `true`  the validation state of all form elements of the form should be reset when the object bound to `@model` argument changes.

`@resetValidationStateOnModelChange` needs to be `false` by default for backward compatibility. We might want to change default value in next major release.

**Comment by rwwagner90** [comment]
@jelhan 

> Not sure if I can follow. I wasn't describing a feature existing today but a potential public API to serve the use case.

Yes, I did not say it existed, I said it sounds reasonable and I agree.

I think `@resetValidationStateOnModelChange` sounds like a good API. Shouldn't it be true always though? If the whole model changes, I am not sure why we would want to keep old validations around.

**Comment by jelhan** [comment]
> I think `@resetValidationStateOnModelChange` sounds like a good API. Shouldn't it be true always though? If the whole model changes, I am not sure why we would want to keep old validations around.

We can't introduce breaking changes. Therefore it must be `false` at least until the next major.

**Comment by rwwagner90** [comment]
@dspigarelli @rstudner perhaps this is something someone in Expel could take on internally? Could be a good way to contribute back 😃

**Comment by jelhan** [comment]
Is this still an issue in recent versions? I would expect that `<BsModal>` and a `<BsForm>` components are teardown when the modal is hidden. But I could be wrong. At least I cannot recall having noticed this bug.
